src/parallel.cpp

```diff
@@ -549,6 +549,10 @@ void HandleBlockMessageThread(CNode *pfrom, const string strCommand, const CBloc
             if (nDoS > 0)
                 dosMan.Misbehaving(pfrom, nDoS);
         }
+
+        // the current fork is bad due to this block so reset the best header to the best fully-validated block
+        // so we can download another fork of headers (and blocks).
+        pindexBestHeader = chainActive.Tip();
```

Comment:
if you really need to do this here maybe you can try:

if (FindMostWorkChain()->nChainWork > chainActive.Tip()->nChainWork)
    pindexBestHeader = FinMostWorkChain();
else
    pindexBestHeader = chainActive.Tip();